### ISSUE TITLE
App Authentication

### ISSUE THREAD

**Issue by Alejandro-HUB** [Alejandro-HUB/Muszilla#14]
#  App Athentication #14 

Currently, we use a non-secure and unorthodox method of authentication for our web API as well as our sessions. 

We are exposing our queries in a string using TSQL.
<img width="1032" alt="image" src="https://user-images.githubusercontent.com/58063303/182518059-5f18004f-6518-411f-ac18-d78d6484d35c.png">

This is bad practice because anyone could perform a database injection against our app. 

## Step 1

Look into Entity Framework's Database Context. You can look into this video from an APP that I made that is using Entity Framework. [Data Context Video](https://1drv.ms/v/s!ApEeJO6L-j83l7IOZcZ472Js9MC0yg)  

Replace only the Endpoints that you have worked on with Data Context for now. That is:

- Login
- Help Page
- Profile Page
- Settings

The essential part here is that using DB context we do not need to perform such queries. Instead, we would use [LINQ](https://docs.microsoft.com/en-us/dotnet/csharp/linq/).

An example of its usage would be having a list of Contacts and you want to find the contact that had a specific ID.

```
Class Contact
{
    ID {get; set;}
    Name {get; set;}
    Email {get; set;}
}

List<Contact> testContact = testContact.Where(x => x.ID == 1).FirstOrDefault();
```

We are running a query against the database to check if the password for that email matches and then storing all of the user's data in the web browser's cache. 

<img width="598" alt="image" src="https://user-images.githubusercontent.com/58063303/182516032-507605f4-66f6-400e-b12b-1569d41d06e0.png">

As you can see in the image above we are using HTTPContext to store the user's data on the web browser's cached data. 

## Step 2

As you can see throughout our home controller we constantly have to get the cached data when we return a view. There is a way to store this data in the user's session without the need to keep re-initializing it throughout the application on each endpoint. Moreover, you can look into how to store the user's data on Models instead of only in the user's cached web session. 

Look into Angular JS and $scope. You can do ng-bind a UI component to a model and its value will be kept without the need to re-initializing it. 

## Step 3 

Another issue here is that when the user logs out and then they get redirected back to the login page if they press the back button on the browser then what I did back then is that I cleared all the cache in the browser and made the UI not display anything the if data in the cache was blank. However, I believe there are better ways to do this.

## How our APP works and how we use endpoints:

For more insight on this, the way our app works is that we call our API endpoints located in our controllers in our JavaScript as a request and we return data back to the JavaScript from the controller. This is already done through our app using ajax.

Example:
When deleting a song I create a request on the client side:
<img width="299" alt="image" src="https://user-images.githubusercontent.com/58063303/182516837-3b0e226e-79bf-4141-8603-07135ba3c548.png">

As you can see the endpoint for this API is /Home/deleteSong and it is a POST. This means we are posting data to our database or executing some sort of action against it. The data sent in the request is the SongToDeleteID and the songName which comes from when the user clicks the delete button in the UI. 

Then if you go inside the HomeController you will see that there is a method called "deleteSong". This method will return a response with whether it was successful in deleting the song or not. 

The point here is after my explanation, while I am hashing the user's password, there are better ways for authentication. 

## Step 4 
Look into API token-based authentication. Like that, only the users that are authenticated can perform API calls to our database.

Example: Using postman I can call one of our endpoints without the need for authentication and delete a song:

<img width="647" alt="image" src="https://user-images.githubusercontent.com/58063303/182520744-02b2a15a-b032-45dc-b969-50d35945ea7c.png">

Looks like even though I had the name wrong, the song with ID **1** in our database got deleted without any authentication. Only logged-in (authenticated) users should be able to do this. 

See? Song with ID **1** got deleted from the DB:

<img width="582" alt="image" src="https://user-images.githubusercontent.com/58063303/182521291-5695586c-a76d-403a-87a7-f912a88b068c.png">

Look into this: 
https://www.youtube.com/watch?v=MltKzQpaltk&t=632s

Basically, before any endpoint calls are made we have to make a new endpoint where we log in and that generates a token. We need to then use that token as authentication in our deleteSongs endpoint so that only if we have a valid token we can delete a song from our database. 

<img width="641" alt="image" src="https://user-images.githubusercontent.com/58063303/182521193-f342668f-3dcc-4ffc-b355-e243e49fb01f.png">
